cluster.go

```diff
@@ -464,48 +464,60 @@ func resolveAddr(addr string) (*net.TCPAddr, error) {
 		}
 	}
 
-	// Attempt to resolve IPv4 and v6 concurrently.
-	addrChan := make(chan *net.TCPAddr, 2)
-	for _, network := range []string{"udp4", "udp6"} {
+	// Attempt to resolve IPv4 and v6 and Unix Domain Sockets concurrently.
+	addrChan := make(chan net.Addr, 3)
+	var i int
+	for _, network := range []string{"udp4", "udp6", "unix"} {
 		network := network
 		go func() {
 			// The unfortunate UDP dialing hack allows having a timeout on address resolution.
 			conn, err := net.DialTimeout(network, addr, 10*time.Second)
 			if err != nil {
 				addrChan <- nil
 			} else {
-				addrChan <- (*net.TCPAddr)(conn.RemoteAddr().(*net.UDPAddr))
+				switch conn.RemoteAddr().(type) {
+				case *net.UDPAddr:
+					addrChan <- (*net.TCPAddr)(conn.RemoteAddr().(*net.UDPAddr))
+				case *net.UnixAddr:
+					addrChan <- conn.RemoteAddr()
+				}
+
 				conn.Close()
 			}
 		}()
 	}
 
-	// Wait for the result of IPv4 and v6 resolution. Use IPv4 if available.
-	tcpaddr := <-addrChan
-	if tcpaddr == nil || len(tcpaddr.IP) != 4 {
-		var timeout <-chan time.Time
-		if tcpaddr != nil {
-			// Don't wait too long if an IPv6 address is known.
-			timeout = time.After(50 * time.Millisecond)
-		}
-		select {
-		case <-timeout:
-		case tcpaddr2 := <-addrChan:
-			if tcpaddr == nil || tcpaddr2 != nil {
-				// It's an IPv4 address or the only known address. Use it.
-				tcpaddr = tcpaddr2
+	// Wait for the result of IPv4, v6 and Unix Domain Socket resolution. Use IPv4 if available.
+outer:
+	for raddr := range addrChan {
+		i++
+
+		switch raddr.(type) {
+		case *net.TCPAddr:
+			tcpaddr := (*net.TCPAddr)(raddr.(*net.TCPAddr))
+			if len(tcpaddr.IP) != 4 {
+				// Don't wait too long if an IPv6 address is known.
+				timeout := time.After(50 * time.Millisecond)
+				<-timeout
```

Comment:
I think the purpose of this timeout has been lost a little here - it blocks the channel receiver loop now, however before it set a timeout that was triggered if a `tcp4` address was not resolved within 50ms.

The `select {}` statement on channels provides [non-blocking](https://gobyexample.com/non-blocking-channel-operations) operations, so before this loop would receive a `tcp6` address and wait for _either_ 50ms before returning the `tcp6` address, or if a `tcp4` address was received before the 50ms deadline that instead. It's to prioritise `tcp4` over `tcp6` addresses (why, I don't know).

The whole `resolveAddr()` func as it exists today is really overcomplicated so it's really not very easy to follow - sorry about that. I think a quick rewrite is in order in general, but for your case you could avoid all this (and provide a faster code path when using `unix`) by checking if it's a unix address at the start of `resolveAddr()` and returning that immediately if it is - no resolving needed!